src/passes/OptimizeInstructions.cpp

```diff
@@ -459,6 +460,50 @@ struct OptimizeInstructions
         }
         // note that both left and right may be consts, but then we let
         // precompute compute the constant result
+      } else if (binary->op == LtSInt32) {
+        if (auto* c = binary->right->dynCast<Const>()) {
+          auto constValue = c->value.geti32();
+          if (constValue == 0) {
+            // this may increase size for compressed binaries
+            // but it could be faster on some runtimes.
+            // Also it canonical form produced by LLVM codegen
+            if (options.shrinkLevel == 0) {
+              // (signed)x < 0 => (unsigned)x >> 31
+              binary->op = ShrUInt32;
+              c->value = Literal(int32_t(31));
+              return binary;
+            }
+          } else if (constValue == int32_t(0x7FFFFFFF)) {
+            // (signed)x < 0x7FFFFFFF => x != 0x7FFFFFFF
+            binary->op = NeInt32;
+            return binary;
+          } else if (constValue == int32_t(0x80000000) &&
+                     !EffectAnalyzer(options, features, binary->left)
+                        .hasSideEffects()) {
+            // (signed)x < 0x80000000 => 0
+            return LiteralUtils::makeZero(Type::i32, *getModule());
+          }
+        }
+      } else if (binary->op == GtSInt32) {
+        if (auto* c = binary->left->dynCast<Const>()) {
+          auto constValue = c->value.geti32();
+          if (constValue == 0) {
+            // 0 > (signed)x => (unsigned)x >> 31
+            binary->op = ShrUInt32;
+            c->value = Literal(int32_t(31));
+            std::swap(binary->left, binary->right);
+            return binary;
+          } else if (constValue == int32_t(0x7FFFFFFF)) {
+            // 0x7FFFFFFF > (signed)x => x != 0x7FFFFFFF
+            binary->op = NeInt32;
+            return binary;
+          } else if (constValue == int32_t(0x80000000) &&
+                     !EffectAnalyzer(options, features, binary->right)
+                        .hasSideEffects()) {
+            // 0x80000000 > (signed)x => 0
+            return LiteralUtils::makeZero(Type::i32, *getModule());
+          }
+        }
```

Comment:
Have you considered canonicalizing the expression to move the constant to the left side (or similar) to halve the number of patterns we need to recognize? It looks like these wouldn't be necessary if the expressions were canonicalized.